backend/remote-state/http/client.go

```diff
@@ -0,0 +1,298 @@
+package http
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	multierror "github.com/hashicorp/go-multierror"
+
+	"github.com/hashicorp/terraform/state"
+	"github.com/hashicorp/terraform/state/remote"
+)
+
+// RemoteClient is used by "state/remote".State to read and write
+// blobs representing state.
+// Implements "state/remote".ClientLocker
+type RemoteClient struct {
+	client *http.Client
+
+	// The fields below are set from configure
+	address       string
+	updateMethod  string
+	lockAddress   string
+	unlockMethod  string
+	lockMethod    string
+	unlockAddress string
+	username      string
+	password      string
+
+	lockID       string
+	jsonLockInfo []byte
+}
+
+// Get blah
+func (c *RemoteClient) Get() (*remote.Payload, error) {
+	// Convert address to type URL
+	addressURL, _ := url.Parse(c.address)
+
+	resp, err := c.httpRequest("GET", addressURL, nil, "get state")
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Handle the common status codes
+	switch resp.StatusCode {
+	case http.StatusOK:
+		// Handled after
+	case http.StatusNoContent:
+		return nil, nil
+	case http.StatusNotFound:
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("Unexpected HTTP response code %d", resp.StatusCode)
+	}
+
+	// Read in the body
+	buf := bytes.NewBuffer(nil)
+	if _, err := io.Copy(buf, resp.Body); err != nil {
+		return nil, fmt.Errorf("Failed to read remote state: %s", err)
+	}
+
+	// Create the payload
+	payload := &remote.Payload{
+		Data: buf.Bytes(),
+	}
+
+	md5 := md5.Sum(buf.Bytes())
+	// If there was no data, then return nil
+	if len(payload.Data) == 0 {
+		return nil, nil
+	}
+
+	// Check for the MD5
+	if raw := resp.Header.Get("Content-MD5"); raw != "" {
+		md5, err := base64.StdEncoding.DecodeString(raw)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"Failed to decode Content-MD5 '%s': %s", raw, err)
+		}
+
+		payload.MD5 = md5
+	} else {
+		// Generate the MD5
+		hash := md5
+		payload.MD5 = hash[:]
+	}
+
+	return payload, nil
+}
+
+// Put state
+func (c *RemoteClient) Put(data []byte) error {
+	// Copy the target URL
+	addressURL, _ := url.Parse(c.address)
+
+	base := *addressURL
+
+	if c.lockID != "" {
+		query := base.Query()
+		query.Set("ID", c.lockID)
+		base.RawQuery = query.Encode()
+	}
+
+	var method = c.updateMethod
+	resp, err := c.httpRequest(method, &base, &data, "upload state")
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Handle the error codes
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusCreated, http.StatusNoContent:
+		return nil
+	default:
+		return fmt.Errorf("HTTP error: %d", resp.StatusCode)
+	}
+}
+
+// Delete state
+func (c *RemoteClient) Delete() error {
+	// Make the request
+	addressURL, _ := url.Parse(c.address)
+	resp, err := c.httpRequest("DELETE", addressURL, nil, "delete state")
```

Comment:
use `http.MethodDelete` constant

Reply:
They are using it this way. Didn't wanted to change it... I guess I can change it to http.MethodGet since there is no clear view of LOCK and UNLOCK. 
https://github.com/hashicorp/terraform/blob/master/state/remote/http.go#L227